File: src/sugtao4423/lod/dialog_onclick/Dialog_favorite.java
package sugtao4423.lod.dialog_onclick;

import twitter4j.Status;
import twitter4j.TwitterException;
import android.app.AlertDialog;
import android.content.Context;
import android.os.AsyncTask;
import android.view.View;
import android.view.View.OnClickListener;
import sugtao4423.lod.ApplicationClass;
import sugtao4423.lod.R;
import sugtao4423.lod.ShowToast;

public class Dialog_favorite implements OnClickListener{

	private Status status;
	private Context context;
	private AlertDialog dialog;

	public Dialog_favorite(Status status, Context context, AlertDialog dialog){
		this.status = status;
		this.context = context;
		this.dialog = dialog;
	}

	@Override
	public void onClick(View v){
		dialog.dismiss();
		new AsyncTask<Void, Void, Boolean>(){
			@Override
			protected Boolean doInBackground(Void... params){
				try{
					((ApplicationClass)context.getApplicationContext()).getTwitter().createFavorite(status.getId());
					return true;
				}catch(TwitterException e){
					return false;
				}
			}

			@Override
			protected void onPostExecute(Boolean result){
				if(result)
					new ShowToast(context.getApplicationContext(), R.string.success_fav);
				else
					new ShowToast(context.getApplicationContext(), R.string.error_fav);
			}
		}.execute();
	}
}

File: app/src/main/java/sugtao4423/lod/dialog_onclick/Dialog_favorite.kt
package sugtao4423.lod.dialog_onclick

import android.content.Context
import android.view.View
import androidx.appcompat.app.AlertDialog
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import sugtao4423.lod.App
import sugtao4423.lod.R
import sugtao4423.lod.utils.showToast
import twitter4j.Status
import twitter4j.TwitterException

class Dialog_favorite(private val status: Status, private val context: Context, private val dialog: AlertDialog) : View.OnClickListener {

    override fun onClick(v: View?) {
        dialog.dismiss()
        if (status.isFavorited) {
            AlertDialog.Builder(context).apply {
                setMessage(R.string.is_unfavorite)
                setNegativeButton(R.string.cancel, null)
                setPositiveButton(R.string.ok) { _, _ ->
                    favorite(false)
                }
                show()
            }
        } else {
            favorite(true)
        }
    }

    private fun favorite(isFavorite: Boolean) {
        CoroutineScope(Dispatchers.Main).launch {
            val result = withContext(Dispatchers.IO) {
                val twitter = (context.applicationContext as App).twitter
                try {
                    if (isFavorite) {
                        twitter.createFavorite(this@Dialog_favorite.status.id)
                    } else {
                        twitter.destroyFavorite(this@Dialog_favorite.status.id)
                    }
                } catch (e: TwitterException) {
                    null
                }
            }
            val toastMessage = when {
                result != null && isFavorite -> R.string.success_favorite
                result != null && !isFavorite -> R.string.success_unfavorite
                result == null && isFavorite -> R.string.error_favorite
                result == null && !isFavorite -> R.string.error_unfavorite
                else -> -1
            }
            context.showToast(toastMessage)
        }
    }

}
